# HG changeset patch
# User Ethan Biery <ebiery@lutron.com>
# Date 1443731593 -7200
#      Thu Oct 01 22:33:13 2015 +0200
# Branch stable
# Node ID 66adbb6e88adee55a85619d2869f26029bbc9784
# Parent  bbe6a846d8ac938b17764dc5175407c2e7714268
textread.m, textscan.m: always remove headerlines args before invoking strread (bug #46080)

diff --git a/scripts/io/textread.m b/scripts/io/textread.m
--- a/scripts/io/textread.m
+++ b/scripts/io/textread.m
@@ -138,21 +138,20 @@ function varargout = textread (filename,
        || ! isnumeric (varargin{headerlines + 1}))
       error ("missing or illegal value for 'headerlines'" );
     endif
     ## Avoid conveying floats to fskipl
     varargin{headerlines + 1} = round (varargin{headerlines + 1});
     ## Beware of zero valued headerline, fskipl would skip to EOF
     if (varargin{headerlines + 1} > 0)
       fskipl (fid, varargin{headerlines + 1});
-      varargin(headerlines:headerlines+1) = [];
-      nargin = nargin - 2;
     elseif (varargin{headerlines + 1} < 0)
       warning ("textread: negative headerline value ignored");
     endif
+    varargin(headerlines:headerlines+1) = [];
   endif
   st_pos = ftell (fid);
 
   ## Read a first file chunk. Rest follows after endofline processing
   [str, count] = fscanf (fid, "%c", BUFLENGTH);
   if (isempty (str) || count < 1)
     warning ("textread: empty file");
     varargout = cell (1, nargout);
diff --git a/scripts/io/textscan.m b/scripts/io/textscan.m
--- a/scripts/io/textscan.m
+++ b/scripts/io/textscan.m
@@ -197,21 +197,21 @@ function [C, position] = textscan (fid, 
          || ! isnumeric (args{headerlines + 1}))
         error ("Missing or illegal value for 'headerlines'" );
       endif
       ## Avoid conveying floats to fskipl
       args{headerlines + 1} = round (args{headerlines + 1});
       if (args{headerlines + 1} > 0)
         ## Beware of zero valued headerline, fskipl would skip to EOF
         fskipl (fid, args{headerlines + 1});
-        args(headerlines:headerlines+1) = [];
         st_pos = ftell (fid);
       elseif (args{headerlines + 1} < 0)
         warning ("textscan.m: negative headerline value ignored");
       endif
+      args(headerlines:headerlines+1) = [];
     endif
     ## Read a first file chunk. Rest follows after endofline processing
     [str, count] = fscanf (fid, "%c", BUFLENGTH);
 
   endif
 
   ## Check for empty result
   if (isempty (str))
